packages/realm-react/src/RealmProvider.tsx

```diff
@@ -41,43 +41,71 @@ type ProviderProps = PartialRealmConfiguration & {
    * instance outside of a component that uses the Realm hooks.
    */
   realmRef?: React.MutableRefObject<Realm | null>;
+  /**
+   * The fallback component to render if the Realm is not open.
+   */
+  fallback?: React.ComponentType<unknown> | React.ReactElement | null | undefined;
   children: React.ReactNode;
-};
+} & PartialRealmConfiguration;
+
+/**
+  Represents the provider returned from using an existing realm at context creation i.e. `createRealmContext(new Realm))`.
+ */
+export type RealmProviderFromRealmFC = React.FC<Pick<RealmProviderProps, "children">>;
+
+/**
+ * Represents the provider returned from using a Realm configuration at context creation i.e. `createRealmContext({schema: []}))`.
+ */
+export type RealmProviderFromConfiguration = React.FC<
+  Pick<RealmProviderProps, "closeOnUnmount" | "realmRef" | "fallback" | "children" | keyof PartialRealmConfiguration>
+>;
+
+/**
+ * Explicitly sets the unpicked properties of a type to never instead of dropping them like in Pick.
+ * Useful for ensuring different prop types are mutually exclusive as React expects the union type
+ * of different prop types to include all the fields.
+ */
+type RestrictivePick<T, K extends keyof T> = Pick<T, K> & { [RestrictedKey in keyof Omit<T, K>]?: never };
+
+/**
+ * Represents properties of a generic RealmProvider where the realm is set and therefore all other
+ * props should be disallowed.
+ */
+export type RealmProviderWithRealmProps = RestrictivePick<RealmProviderProps, "realm" | "children">;
+
+/*
+ * Represents properties of a generic RealmProvider where Realm configuration props are used and no realm is set
+ */
+export type RealmProviderWithConfigurationProps = RestrictivePick<
+  RealmProviderProps,
+  "closeOnUnmount" | "realmRef" | "fallback" | "children" | keyof PartialRealmConfiguration
+>;
+
+/**
+ * Represents the provider returned from creating context with no arguments (including the default context).
+ * Supports either passing a `realm` as a property or the schema configuration.
+ */
+export type DynamicRealmProvider = React.FC<RealmProviderWithRealmProps | RealmProviderWithConfigurationProps>;
+
+export function createRealmProviderFromRealm(
+  realm: Realm | null,
```

Comment:
It looks like all the callers of this function passes a `Realm` instance?
```suggestion
  realm: Realm,
```